File: app/src/main/java/com/kimikevin/eatsplorer/model/repository/RestaurantRepository.java
package com.kimikevin.eatsplorer.model.repository;

import android.util.Log;

import androidx.annotation.NonNull;
import androidx.lifecycle.MutableLiveData;

import com.kimikevin.eatsplorer.BuildConfig;
import com.kimikevin.eatsplorer.model.entity.NearbySearchRequest;
import com.kimikevin.eatsplorer.model.entity.NearbySearchResponse;
import com.kimikevin.eatsplorer.model.entity.PlaceDetailsResponse;
import com.kimikevin.eatsplorer.model.entity.PlacesService;
import com.kimikevin.eatsplorer.model.entity.Restaurant;
import com.kimikevin.eatsplorer.model.mapper.RestaurantMapper;

import java.util.List;

import retrofit2.Call;
import retrofit2.Callback;
import retrofit2.Response;

/**
 * Repository class that abstracts the data source (Google Places API).
 */
public class RestaurantRepository {
    private static final String TAG = "RestaurantRepo";
    private static final String API_KEY = BuildConfig.GOOGLE_MAPS_API_KEY;
    
    private final PlacesService apiService;
    private static RestaurantRepository instance;

    private RestaurantRepository() {
        apiService = RetrofitClient.getClient().create(PlacesService.class);
    }

    public static synchronized RestaurantRepository getInstance() {
        if (instance == null) {
            instance = new RestaurantRepository();
        }
        return instance;
    }

    /**
     * Search for nearby restaurants based on latitude and longitude.
     */
    public void searchNearby(double lat, double lng, MutableLiveData<List<Restaurant>> liveData, MutableLiveData<String> errorData) {
        NearbySearchRequest requestBody = new NearbySearchRequest(lat, lng, 5000);
        apiService.searchNearby(API_KEY, PlacesService.LIST_FIELD_MASK, requestBody)
                .enqueue(new Callback<>() {
                    @Override
                    public void onResponse(@NonNull Call<NearbySearchResponse> call, @NonNull Response<NearbySearchResponse> response) {
                        if (response.isSuccessful() && response.body() != null) {
                            List<Restaurant> cleanList = RestaurantMapper.mapToDomain(response.body().getPlaces());
                            liveData.postValue(cleanList);
                        } else {
                            handleError(response.code(), response.message(), errorData);
                        }
                    }

                    @Override
                    public void onFailure(@NonNull Call<NearbySearchResponse> call, @NonNull Throwable t) {
                        handleFailure(t, errorData);
                    }
                });
    }

    /**
     * Fetch detailed information for a specific restaurant.
     */
    public void getPlaceDetails(String placeId, MutableLiveData<PlaceDetailsResponse> liveData, MutableLiveData<String> errorData) {
        apiService.getPlaceDetails(placeId, API_KEY, PlacesService.DETAILS_FIELD_MASK, "en")
                .enqueue(new Callback<>() {
                    @Override
                    public void onResponse(@NonNull Call<PlaceDetailsResponse> call, @NonNull Response<PlaceDetailsResponse> response) {
                        if (response.isSuccessful() && response.body() != null) {
                            liveData.postValue(response.body());
                        } else {
                            handleError(response.code(), response.message(), errorData);
                        }
                    }

                    @Override
                    public void onFailure(@NonNull Call<PlaceDetailsResponse> call, @NonNull Throwable t) {
                        handleFailure(t, errorData);
                    }
                });
    }

    private void handleError(int code, String message, MutableLiveData<String> errorData) {
        String errorMessage = "API Error (" + code + "): " + message;
        Log.e(TAG, errorMessage);
        errorData.postValue(errorMessage);
    }

    private void handleFailure(Throwable t, MutableLiveData<String> errorData) {
        String failureMessage = "Network Failure: " + t.getMessage();
        Log.e(TAG, failureMessage, t);
        errorData.postValue(failureMessage);
    }
}

File: app/src/main/java/com/kimikevin/eatsplorer/model/repository/RestaurantRepository.kt
package com.kimikevin.eatsplorer.model.repository

import android.util.Log
import com.kimikevin.eatsplorer.BuildConfig
import com.kimikevin.eatsplorer.model.entity.*
import com.kimikevin.eatsplorer.model.mapper.RestaurantMapper
import okhttp3.OkHttpClient
import okhttp3.logging.HttpLoggingInterceptor
import retrofit2.Retrofit
import retrofit2.converter.gson.GsonConverterFactory

class RestaurantRepository private constructor() {
    private val apiService: PlacesService

    init {
        val logging = HttpLoggingInterceptor().apply {
            level = HttpLoggingInterceptor.Level.BODY
        }

        val client = OkHttpClient.Builder()
            .addInterceptor(logging)
            .build()

        val retrofit = Retrofit.Builder()
            .baseUrl(BASE_URL)
            .client(client)
            .addConverterFactory(GsonConverterFactory.create())
            .build()

        apiService = retrofit.create(PlacesService::class.java)
    }

    suspend fun searchNearby(lat: Double, lng: Double): Result<List<Restaurant>> {
        return try {
            val requestBody = NearbySearchRequest(lat, lng, 5000.0)
            val response = apiService.searchNearby(API_KEY, PlacesService.LIST_FIELD_MASK, requestBody)
            if (response.isSuccessful) {
                val body = response.body()
                val cleanList = RestaurantMapper.mapToDomain(body?.places)
                Result.success(cleanList)
            } else {
                val errorMsg = "Error: ${response.code()} ${response.message()}"
                Log.e(TAG, errorMsg)
                Result.failure(Exception(errorMsg))
            }
        } catch (e: Exception) {
            Log.e(TAG, "Network Failure: ${e.message}")
            Result.failure(e)
        }
    }

    suspend fun getPlaceDetails(placeId: String): Result<PlaceDetailsResponse> {
        return try {
            val response = apiService.getPlaceDetails(placeId, API_KEY, PlacesService.DETAILS_FIELD_MASK, "en")
            if (response.isSuccessful && response.body() != null) {
                Result.success(response.body()!!)
            } else {
                val errorMsg = "Error fetching details: ${response.message()}"
                Log.e(TAG, errorMsg)
                Result.failure(Exception(errorMsg))
            }
        } catch (e: Exception) {
            Log.e(TAG, "Network Failure: ${e.message}")
            Result.failure(e)
        }
    }

    companion object {
        private const val TAG = "RestaurantRepo"
        private const val BASE_URL = "https://places.googleapis.com/"
        private val API_KEY = BuildConfig.GMP_KEY

        @Volatile
        private var instance: RestaurantRepository? = null

        fun getInstance(): RestaurantRepository {
            return instance ?: synchronized(this) {
                instance ?: RestaurantRepository().also { instance = it }
            }
        }
    }
}
